faunadb-java-dsl/src/main/java/com/faunadb/client/types/Encoder.java

```diff
@@ -0,0 +1,130 @@
+package com.faunadb.client.types;
+
+import com.faunadb.client.types.Properties.Property;
+import com.faunadb.client.types.Value.ArrayV;
+import com.faunadb.client.types.Value.ObjectV;
+
+import java.lang.reflect.Array;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Stack;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static com.google.common.collect.Lists.newArrayListWithCapacity;
+import static com.google.common.collect.Maps.newHashMap;
+import static java.lang.String.format;
+
+/**
+ * FaunaDB object to {@link Value} encoder.
+ */
+public final class Encoder {
+  private Encoder() {
+  }
+
+  private final static Stack<Object> stack = new Stack<>();
```

Comment:
It cannot be like that, if we create the encoder inside the codec, when we gonna encode a field of a nested class it will instantiate a new encoder, so the stack object would be empty so we wouldn't be able to detect recursive references. To it work properly we would have to instantiate the encoder in the beginning of the encoding process and use the very same encoder in the whole process